### ISSUE TITLE
[BUG] 

### ISSUE THREAD

**Issue by sck-at-ucy** [ml-explore/mlx#1259]
**Describe the bug**
I am trying to set up my code so that it periodically saves the model and optimizer states (with a predefined checkpoint frequency) so that if desired I can go back to reload the model and optimizer states and continue the training from any of the saved checkpoints. I am using `@compile` with the train_step function. If I simply save and reload without actually quitting the code the training seems to continue as expected based on the loss decrease trend. However, if I actually quit the code and try to reload in a fresh run there seems to be jump in loss, so it seems that I am not exactly reproducing the state of the training. Below are relevant code snippets.   

**To Reproduce**

Include code snippet
```python
def save_model_and_optimizer(config, current_epoch, dir_path, model_base_file_name, optimizer_base_file_name, hyper_base_file_name):
    if not os.path.exists(dir_path):
        os.makedirs(dir_path)

    model_file_name = f"{model_base_file_name}_epoch_{current_epoch}.pkl"
    optimizer_file_name = f"{optimizer_base_file_name}_epoch_{current_epoch}.pkl"
    random_state_file_name = f"random_state_epoch_{current_epoch}.pkl"
    config_file_name = f"{hyper_base_file_name}_epoch_{current_epoch}.json"

    # Save model state (parameters only)
    model_file_path = os.path.join(dir_path, model_file_name)
    with open(model_file_path, 'wb') as f:
        pickle.dump(model.parameters(), f)

    # Save optimizer state
    optimizer_file_path = os.path.join(dir_path, optimizer_file_name)
    with open(optimizer_file_path, 'wb') as f:
        pickle.dump(optimizer.state, f)

    # Save random state
    random_state_path = os.path.join(dir_path, random_state_file_name)
    with open(random_state_path, 'wb') as f:
        pickle.dump(mx.random.state, f)

    # Save training configuration
    config['current_epoch'] = current_epoch
    hyper_file_path = os.path.join(dir_path, config_file_name)
    with open(hyper_file_path, 'w') as json_file:
        json.dump(config, json_file, indent=4)

    print(f"Model, optimizer, random state, and configuration saved at epoch {current_epoch}.")

def load_model_and_optimizer(model, optimizer, dir_path, model_base_file_name, optimizer_base_file_name, hyper_base_file_name, checkpoint_epoch):
    model_file_name = f"{model_base_file_name}_epoch_{checkpoint_epoch}.pkl"
    optimizer_file_name = f"{optimizer_base_file_name}_epoch_{checkpoint_epoch}.pkl"
    random_state_file_name = f"random_state_epoch_{checkpoint_epoch}.pkl"
    config_file_name = f"{hyper_base_file_name}_epoch_{checkpoint_epoch}.json"

    model_file_path = os.path.join(dir_path, model_file_name)
    optimizer_file_path = os.path.join(dir_path, optimizer_file_name)
    random_state_file_path = os.path.join(dir_path, random_state_file_name)
    config_file_path = os.path.join(dir_path, config_file_name)

    if os.path.exists(model_file_path) and os.path.exists(optimizer_file_path) and os.path.exists(random_state_file_path) and os.path.exists(config_file_path):
        # Load model state (parameters only)
        with open(model_file_path, 'rb') as f:
            parameters = pickle.load(f)
            model.update(parameters)

        # Load optimizer state
        with open(optimizer_file_path, 'rb') as f:
            optimizer.state = pickle.load(f)

        # Load random state
        with open(random_state_file_path, 'rb') as f:
            mx.random.state = pickle.load(f)

        # Load training configuration
        with open(config_file_path, 'r') as json_file:
            config = json.load(json_file)

        start_epoch = config.get('current_epoch', 0)
        print(f"Model, optimizer, random state, and configuration loaded from {dir_path} at epoch {checkpoint_epoch}.")
    else:
        start_epoch = 0
        print(f"No saved model found at epoch {checkpoint_epoch}. Starting training from scratch.")

    return start_epoch, config


    @partial(mx.compile, inputs=state, outputs=state)
    def train_step(src, target, src_alphas, src_dts, dx, dy):
        loss_and_grad_fn = nn.value_and_grad(model, loss_fn_2D)
        loss, grads = loss_and_grad_fn(model, src, target, src_alphas, src_dts, dx, dy)
        optimizer.update(model, grads)
        return loss, grads

    @partial(mx.compile, inputs=state, outputs=state)
    def evaluate_step(src, target, src_alphas, src_dts, dx, dy):
        loss = loss_fn_2D(model, src, target, src_alphas, src_dts, dx, dy)
        return loss
```

**Expected behavior**
During the initial run the loss history is as follows 

```
start_epoch: 0 epochs: 10 learing_rate: array(0, dtype=float32)
Epoch 1, Training Loss: 12.25573897861299, Validation Loss: 12.527009956240654
Epoch 2, Training Loss: 12.255739007620585, Validation Loss: 12.527009927630424
Model, optimizer, random state, and configuration saved at epoch 2.
Epoch 3, Training Loss: 0.250323773617191, Validation Loss: 0.09001006038859487
Epoch 4, Training Loss: 0.05461768370742599, Validation Loss: 0.03122760124504566
Model, optimizer, random state, and configuration saved at epoch 4.
Epoch 5, Training Loss: 0.03391449196275235, Validation Loss: 0.014082754662958906
```

I try to reload from the saved state at epoch 4 and expect after one more epoch of training to get a loss in the range of Epoch 5 in the initial training history. Instead I get 

```
Model, optimizer, random state, and configuration loaded from /Users/m2/PycharmProjects/pythonProject_StreamLit/Transformer_save at epoch 4.
************************ Hostname: PaloAlto Starting Training *********************** 
start_epoch: 4 epochs: 10 learing_rate: array(0.0001, dtype=float32)
Epoch 5, Training Loss: 0.31006237406904497, Validation Loss: 0.06310294904746115
```

**Desktop (please complete the following information):**
 - OS Version: [e.g. MacOS 14.1.2]
 - Version [e.g. 0.7.0]

**Additional context**

**Comment by awni** [comment]
Hmm, from a first pass it looks like you should be saving everything you need. I can try running it to see in more detail if you share the driving code (the train loop / data bits).

**Comment by sck-at-ucy** [comment]
Thank you, I will do some cleaning and share. I am using some random shuffling in the data generation so to make sure that I am using exactly the same datasets in the initial and reloaded runs, during the initial run I save the datasets after they are generated and reload them when continuing from a checkpoint. This was an obvious potential source of the problem, but it does not seem to be what is causing the issue.

**Comment by sck-at-ucy** [comment]
Thanks for the debug suggestion it helps some.  Of course when I reload in a fresh run from a saved checkpoint the comparison will fail as the model initialized before the reload will be in a random state and will not match the saved state. But I have implemented the comparison for the case when I save and reload back live every few epochs during the training and the comparison there seems to check (as far as I can tell, I still have some doubts if I am checking everything as thoroughly as I need) but in the live case of course we know it works, since the save-and-reload does not affect the loss trend. As you said this sounds like something is either not included in the save or is not properly loaded in the state but it does not affect the live case because that state is already correct in that part but it affects the fresh reload because then that part of the state is not properly set. I am busy with admin duties, It might take me a day or so to clean up the code so I can pass you a version that makes sense. 

Thank you 🙏🏻😊 for offering to help debug this!!

**Comment by sck-at-ucy** [comment]
Code shared by direct email.

**Comment by awni** [comment]
The issue seems to be the optimizer state is not getting set properly:

After you load the state from the file, switch the two lines which init the optimizer and then set the state (so it's like so):

```python
        optimizer.init(model.trainable_parameters())
        optimizer.state = loaded_optimizer_state
```

The other thing is I would remove the bit about the warmup which I think might also mess up the state, e.g. the stuff after this line:

```python
    if start_epoch > 1:   # if this is a reload run, do two warm-up iterations first
```

On our side, I think there is a bit of a gotcha with loading and saving the optimizer. If you don't `init` it before updating the state then you will have issues. We should probably fix that so you can simply set the state without needing to initialize it.

**Comment by sck-at-ucy** [comment]
**Thank you! I confirm that implementing these changes solved the issue I was facing ✅**

I notice that in the example you have included in the updated docs you save and load states using  

```python
   state = tree_flatten(optimizer.state)
   mx.save_safetensors("optimizer.safetensors", dict(state))
   # Later on, for example when loading from a checkpoint,
   # recreate the optimizer and load the state
   optimizer = optim.Adam(learning_rate=1e-2)
   state = tree_unflatten(list(mx.load("optimizer.safetensors").
   
```
   
whereas I was using `pickle`. I tried switching but I ran into some issues, it seems to me that the `tree_flatten/unflatten` utility is being more selective as to which keys are saved whereas pickle is dumping everything? For example, it seems that  `tree_flatten/unflatten` omits keys if they correspond to` None/{}` values?  I did not look further into this yet, but the behavior is somewhat different and requires some to implement correctly, for example when I check for matches between current and loaded states. The good news is that the current implementation I have with `pickle` works well after the changes you suggested.

**Comment by awni** [comment]
>  I confirm that implementing these changes solved the issue I was facing 

Great!

> For example, it seems that tree_flatten/unflatten omits keys if they correspond to None/{} values?

I think it should be ok if the empty nodes from the state are left out. Even though you will not get the identical state before and after the load. You will get all the state you need to resume where you left off. But indeed Pickling is also quite reasonable.
